### PR TITLE
Secure gateway:

### PR DESCRIPTION
**Pull request by mtrippled**
This is designed for use by proxies in front of rippled. Configured IPs
can forward identifying user data in HTTP headers, including
user name and origin IP. If the user name exists, then resource limits
are lifted for that session. However, administrative commands are still
reserved only for administrative sessions.

### CODE REVIEW

File: src/ripple/resource/impl/Logic.h [hunk 1]
```diff
@@ -189,42 +194,11 @@ class Logic
         }
 
         m_journal.debug <<
-            "New admin endpoint " << *entry;
+            "New unlimited endpoint " << *entry;
 
         return Consumer (*this, *entry);
     }
 
-    Entry& elevateToAdminEndpoint (Entry& prior, std::string const& name)
```
Comment (vinniefalco) [inline src/ripple/resource/impl/Logic.h]: Why was this removed?

File: src/ripple/rpc/handlers/AccountTx.cpp [hunk 2]
```diff
@@ -129,7 +129,7 @@ Json::Value doAccountTx (RPC::Context& context)
         {
             auto txns = context.netOps.getTxsAccountB (
                 *account, uLedgerMin, uLedgerMax, bForward, resumeToken, limit,
-                context.role == Role::ADMIN);
+                unlimited (context.role));
```
Comment (vinniefalco) [inline src/ripple/rpc/handlers/AccountTx.cpp]: Consider `is_unlimited`

File: src/ripple/rpc/impl/RPCHandler.cpp [hunk 3]
```diff
@@ -232,7 +232,26 @@ Status doCommand (
     }
 
     if (auto method = handler->valueMethod_)
-        return callMethod (context, method, handler->name_, result);
+    {
+        if (context.user.size() || context.forwardedFor.size())
```
Comment (vinniefalco) [inline src/ripple/rpc/impl/RPCHandler.cpp]: Consider `! context.user.empty() || ! context.forwardedFor.empty()`

File: src/ripple/server/Port.h [hunk 4]
```diff
@@ -107,6 +108,13 @@ operator<< (std::ostream& os, Port const& p)
             os << ip.to_string () << ", ";
     }
 
+    if (! p.secure_gateway_ip.empty ())
```
Comment (vinniefalco) [inline src/ripple/server/Port.h]: This routine is getting big for an inline, consider moving the definition to `ripple/server/impl/Port.cpp`

File: src/ripple/server/Role.h [hunk 5]
```diff
@@ -46,12 +52,27 @@ enum class Role
...
+ */
+bool
+unlimited (Role const& role);
+
+/**
+ * If the HTTP header X-User exists with a non-empty value was passed by an IP
+ * configured as secure_gateway, then the user can be positively identified.
+ */
+bool
+identified (HTTP::Port const& port, beast::IP::Address const& remoteIp,
```
Comment (vinniefalco) [inline src/ripple/server/Role.h]: Consider `is_identified`

**Comment by vinniefalco** [comment]
Could use some minor style improvements but its good :+1:

**Comment by mtrippled** [comment]
unittest and npm test both worked locally for me. I've restarted the CI, assuming the errors are intermittent and localized to Travis.

File: src/ripple/resource/impl/Logic.h (hunk 1 above)
Comment (mtrippled) [inline src/ripple/resource/impl/Logic.h]: It's been unused for 25 months: https://github.com/ripple/rippled/commit/a1b487c5128e6531295d6089bef474a391256ce4

**Comment by mtrippled** [comment]
@vinniefalco @nbougalis Your comments have been addressed.

**Comment by vinniefalco** [comment]
Still :+1:

**Comment by mtrippled** [comment]
@vinniefalco @nbougalis @miguelportilla I just added the capability to check the port & role privileges through the ping command.

**Comment by mtrippled** [comment]
@miguelportilla I just push the fixes you suggested.

**Comment by miguelportilla** [comment]
:+1:  :rabbit:

**Comment by nbougalis** [comment]
:+1:
